src/bin/repl.js

```diff
@@ -0,0 +1,231 @@
+#!/usr/bin/env node
+
+const readline = require("readline");
+const { createReadStream } = require("fs");
+
+const { _debug } = require("../index");
+const { createCompiledModule } = require("../compiler/compile/module");
+const { Instance } = require("../interpreter");
+const partialEvaluation = require("../interpreter/partial-evaluation");
+const { Memory } = require("../interpreter/runtime/values/memory");
+const { createAllocator } = require("../interpreter/kernel/memory");
+
+/**
+ * Assert helpers
+ */
+
+// assert module is invalid with given failure string
+// ( assert_invalid <module> <failure> )
+function assert_invalid(node) {
+  const [module, expected] = node.args;
+
+  try {
+    createInstanceFromAst(module);
+
+    assert(false, "module is valid, expected invalid");
+  } catch (err) {
+    assert(err.message === expected);
+  }
+}
+
+// assert action has expected results
+// ( assert_return <action> <expr>* )
+
+// action:
+//   ( invoke <name>? <string> <expr>* )        ;; invoke function export
+//   ( get <name>? <string> )                   ;; get global export
+
+function assert_return(node) {
+  const [action, ...exprs] = node.args;
+
+  if (action.type === "Instr" && action.id === "invoke") {
+    const actualRes = invoke(action);
+    const expectedRes = partialEvaluation
+      .evaluate(allocator, exprs)
+      .value.toNumber();
+
+    assert(
+      actualRes === expectedRes,
+      `expected ${expectedRes}, ${actualRes} given`
+    );
+  } else {
+    throw new Error("Unsupported action in assert_return: " + action.id);
+  }
+}
+
+// invoke function export
+// ( invoke <name>? <string> <expr>* )
+function invoke(node) {
+  const [name, ...exprs] = node.args;
+
+  // find export in instantiated module
+
+  const module = instantiatedModules.find(
+    ({ exports }) => exports[name.value] !== undefined
+  );
+
+  assert(module !== undefined, `Module with export "${name.value}" not found`);
+
+  const args = exprs.map(expr => {
+    const res = partialEvaluation.evaluate(allocator, [expr]);
+    return res.value.toNumber();
```

Comment:
For i64 values `toNumber` will result in lost precision. It might be worth adding a `toString` method to the number interface and using it here.

Reply:
Yes you are right, I assume the spec tests are using i64 in function signatures (which is fine since no JS boundaries are crossed).

I will use the string representation, it's just for comparison anyway, thanks for the hint!